akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -45,18 +52,27 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with StageLoggi
         }
     }
   def handleStreamEvent(e: StreamFrameEvent): Unit = {
-    val newState = streamFor(e.streamId).handle(e)
-    if (newState == Closed) incomingStreams -= e.streamId
-    else incomingStreams += e.streamId -> newState
+    updateState(e.streamId, streamFor(e.streamId).handle(e))
+  }
+  // Called by the outgoing stream multiplexer when that side of the stream is ended.
+  def handleOutgoingEnded(streamId: Int): Unit = {
+    updateState(streamId, streamFor(streamId).handleOutgoingEnded())
+  }
+  def updateState(streamId: Int, newState: IncomingStreamState): Unit = {
+    if (newState == Closed) incomingStreams -= streamId
```

Comment:
This like looks like the overall fix. I can confirm from my tests that it's able to GC all of the Outstreams and incoming calls.